language/stdlib/modules/libra_system.move

```diff
@@ -70,118 +103,251 @@ module LibraSystem {
+    public fun get_validator_info(addr: address): ValidatorInfo {
+        let validator_set = get_validator_set();
+        let validator_index_vec = get_validator_index_(&validator_set.validators, addr);
+        Transaction::assert(!Vector::is_empty(&validator_index_vec), 19);
+
+        *Vector::borrow(&validator_set.validators, *Vector::borrow(&validator_index_vec, 0))
     }
 
     // Return the size of the current validator set
     public fun validator_set_size(): u64 {
         Vector::length(&get_validator_set().validators)
     }
 
-   fun is_validator_(addr: &address, validators_vec_ref: &vector<ValidatorInfo>): bool {
-        let size = Vector::length(validators_vec_ref);
-        if (size == 0) {
-            return false
+    // This function is used in transaction_fee.move to distribute transaction fees among validators
+    public fun get_ith_validator_address(i: u64): address {
+        Vector::borrow(&get_validator_set().validators, i).addr
+    }
+
+    // Private helper functions
+
+    // The Association, the VM and the validator operator from the current validator set
+    // are authorized to update the set of validator infos
+    fun is_sender_authorized_(): bool {
+        // succeed fast
+        if (Transaction::sender() == 0xA550C18 || Transaction::sender() == 0x0) {
+            return true
         };
+        let validators = &get_validator_set().validators;
+        // scan the validators to find a match
+        let size = Vector::length(validators);
+        // always true: size > 3 (see remove_validator code)
 
         let i = 0;
-        // this is only needed to make the bytecode verifier happy
-        let validator_info_ref = Vector::borrow(validators_vec_ref, i);
+        if (ValidatorConfig::get_validator_operator_account(Vector::borrow(validators, i).addr) ==
+            Transaction::sender()) {
+            return true
+        };
         loop {
-            if (&validator_info_ref.addr == addr) {
+            i = i + 1;
+            if (i >= size) { break };
+            if (ValidatorConfig::get_validator_operator_account(Vector::borrow(validators,
+                                                                               i).addr) ==
+                Transaction::sender()) {
                 return true
             };
+        };
+        return false
+    }
+
+    // Emit a reconfiguration event. This function will be invoked by the genesis
+    // to generate the very first reconfiguration event.
+    // Get the index of the validator by address in the `validators` vector
+    fun get_validator_index_(validators: &vector<ValidatorInfo>, addr: address): vector<u64> {
+        let size = Vector::length(validators);
+        let result: vector<u64> = Vector::empty();
+        if (size == 0) {
+            return result
+        };
+
+        let i = 0;
+        let validator_info_ref = Vector::borrow(validators, i);
+        loop {
+            if (validator_info_ref.addr == addr) {
```

Comment:
When can there be a duplicated address? Is it because of the validator delegation?

Reply:
There can not be duplicate addresses in the ValidatorSet (`LibraSystem::add_validator` asserts that).
The validators' delegates (= validators' operators) are allowed to call `LibraSystem::update_and_reconfigure`, but the ValidatorSet contains a list of unique validators' addresses themselves (not addresses of their delegates).

Reply:
If there can't be duplicated address, why do we need to return a `vector<u64>`?